README updates for ilab 0.18.0 (#2062)
instructlab/instructlab: README.md

   #### Install with Nvidia CUDA

>      For the best CUDA experience, installing VLLM is necessary to serve Safetensor format models.

Reviewer: vLLM? Safetensors?

   If this is your first time running InstructLab, it's best to start with `ilab config init` to create the environment.

   Options:
   --config PATH  Path to a configuration file.  [default:

>                   /home/cdoern/.config/instructlab/config.yaml]

Reviewer: up to you, but maybe better replace particular user names with something more generic.

Author: true, will do that

Reviewer 2: I usually add something like <user-name>

   -v, --verbose  Enable debug logging (repeat for even more verbosity)
   --version      Show the version and exit.
   --help         Show this message and exit.

   Commands:

>    config    Command Group for Interacting with the Config of InstructLab.

Reviewer 3: The capitalization here is really strange, why are words like group, interacting, config, etc. capitalized, while others like 'command', 'related', 'generated', etc. are not? 

If we are adjusting the capitalization in this PR then we should normalize all of the sentences so that the capitalization is in line with standard English rules.

Reviewer 4: I agree, but what's important here is that it matches what's in the CLI

Reviewer 3: Ack, it sounds like this is something we'll want to handle in a separate PR then post-0.18.0. 

I've created an issue for it here: https://github.com/instructlab/instructlab/issues/2064

   Path to taxonomy repo [taxonomy]: <ENTER>
   `taxonomy` seems to not exists or is empty. Should I clone https://github.com/instructlab/taxonomy.git for you? [y/N]: y
   Cloning https://github.com/instructlab/taxonomy.git...
   ```

   `ilab` will use the default configuration file unless otherwise specified. You can override this behavior with the `--config` parameter for any `ilab` command.

> 4. When prompted, provide the path to your default model. Otherwise, we will use the default of merlinite.

Reviewer 4: ```suggestion
4. When prompted, provide the path to your default model. Otherwise, the default of a quantized [Merlinite](https://huggingface.co/instructlab/merlinite-7b-lab-GGUF) model will be used - you can download this model with `ilab model download` (see below).
```

   ```shell
   (venv) $ ilab config init
   Welcome to InstructLab CLI. This guide will help you set up your environment.
   Please provide the following values to initiate the environment [press Enter for defaults]:
   Path to taxonomy repo [taxonomy]: <ENTER>
   `taxonomy` seems to not exists or is empty. Should I clone https://github.com/instructlab/taxonomy.git for you? [y/N]: y
   Cloning https://github.com/instructlab/taxonomy.git...
   Path to your model [/home/user/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf]: <ENTER>
   ```

> 5. When prompted, please choose a train profile. Train profiles are GPU specific profiles that enable accelerated training behavior. If you are on MacOS or a Linux machine without a dedicated GPU, please choose `No Profile (CPU-Only)` by hitting Enter.

Reviewer 4: ```suggestion
5. When prompted, please choose a train profile. Train profiles are GPU specific profiles that enable accelerated training behavior. If you are on MacOS or a Linux machine without a dedicated GPU, please choose `No Profile (CPU-Only)` by hitting Enter. There are various flags you can utilize with individual `ilab` commands that will allow you to utilize your GPU if applicable.
```

> 5. When promoted, please choose a train profile. Train profiles are GPU specific profiles that enable accelerated training behavior. If you are on MacOS or a Linux machine without a dedicated GPU, please choose `No Profile (CPU-Only)` by hitting enter.

Reviewer 3: ```suggestion
5. When prompted, please choose a train profile. Train profiles are GPU specific profiles that enable accelerated training behavior. If you are on MacOS or a Linux machine without a dedicated GPU, please choose `No Profile (CPU-Only)` by hitting enter.
```

> 5. When prompted, please choose a train profile. Train profiles are GPU specific profiles that enable accelerated training behavior. If you are on MacOS or a Linux machine without a dedicated GPU, please choose `No Profile (CPU-Only)` by hitting enter.

Reviewer 5: Is an M1/M2/M3 with Metal considered a dedicated GPU? I'm pretty sure there's a GPU in that machine.

> 5. When prompted, please choose a train profile. Train profiles are GPU specific profiles that enable accelerated training behavior. If you are on MacOS or a Linux machine without a dedicated GPU, please choose `No Profile (CPU-Only)` by hitting enter.

Reviewer: `hitting Enter` (capital `E`)

   ```shell
   Welcome to InstructLab CLI. This guide will help you to setup your environment.
   Please provide the following values to initiate the environment [press Enter for defaults]:
   Path to taxonomy repo [/home/cdoern/.local/share/instructlab/taxonomy]: 
   Path to your model [/home/cdoern/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf]: 
   Generating `/home/cdoern/.config/instructlab/config.yaml`...
   Please choose a train profile to use:
   [0] No profile (CPU-only)
   [1] A100_H100_x2.yaml
   [2] A100_H100_x4.yaml
   [3] A100_H100_x8.yaml
   [4] L40_x4.yaml
   [5] L40_x8.yaml
   [6] L4_x8.yaml
   Enter the number of your choice [hit enter for the default CPU-only profile] [0]:  
   Using default CPU-only train profile.
   Initialization completed successfully, you're ready to start using `ilab`. Enjoy!
   ```

>    The GPU profiles are listed by GPU type and amount. If you happen to have a GPU configuration with a similar amount of vram as any of the above configs, feel free to try them out!

Reviewer: GPU type and amount -> number? (I think the former is for "mass nouns").

Reviewer: VRAM?

Reviewer: `above configs` -> maybe `profiles`?

Author: did all of these!

### `ilab` directory layout after initializing your system

> After running `ilab config init` your directories will look like the following on a linux system:

Reviewer 5: ```suggestion
After running `ilab config init` your directories will look like the following on a Linux system:
```

> After running `ilab config init` your directories will look like the following on a linux system:

Reviewer: Linux (capital L)

```shell
├─ ~/.cache/instructlab/models/ (1)
├─ ~/.local/share/instructlab/datasets (2)
├─ ~/.local/share/instructlab/taxonomy (3)
├─ ~/.local/share/instructlab/checkpoints (4)
```

>  1. ~/.cache/instructlab/models/: Contains all downloaded large language models, including the saved output of ones you generate with ilab.
>  2. ~/.local/share/instructlab/datasets/: Contains data output from the SDG phase, built on modifications to the taxonomy repository.
>  3. ~/.local/share/instructlab/taxonomy/: Contains the skill and knowledge data.
>  4. ~/.local/share/instructlab/checkpoints/: Contains the output of the training process

Reviewer 4: ```suggestion
 1. `~/.cache/instructlab/models/`: Contains all downloaded large language models, including the saved output of ones you generate with ilab.
 2. `~/.local/share/instructlab/datasets/`: Contains data output from the SDG phase, built on modifications to the taxonomy repository.
 3. `~/.local/share/instructlab/taxonomy/`: Contains the skill and knowledge data.
 4. `~/.local/share/instructlab/checkpoints/`: Contains the output of the training process
```

>  On MacOS these directories will be under `Library/Application Support/instructlab`. This directory setup is temporary in 0.18.0 and will mimic the Linux paths in future releases. The models directory will be under `Library/Caches/instructlab`.

Reviewer 4: ```suggestion
 On MacOS, these directories will be under `Library/Application Support/instructlab`. This directory setup is temporary in 0.18.0 and will mimic the Linux paths in future releases. The models directory will be under `Library/Caches/instructlab`.
```

  HF_TOKEN=<YOUR HUGGINGFACE TOKEN GOES HERE> ilab model download --repository=TheBloke/Mixtral-8x7B-Instruct-v0.1-GGUF --filename=mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf
  ```

>   #### Downloading an entire Hugging Face repository (Safetensor Model)

Reviewer: Safetensors (I think that's the correct name of the format?) Same elsewhere.

- Specify repository, and a Hugging Face token if necessary. For example:

  ```shell
  HF_TOKEN=<YOUR HUGGINGFACE TOKEN GOES HERE> ilab model download --repository=instructlab/granite-7b-lab
  ```

>   These types of models are useful for GPU enabled systems or anyone looking to serve a model using VLLM. Instructlab provides Safetensor versions of our granite models on HuggingFace.

Reviewer 4: ```suggestion
  These types of models are useful for GPU-enabled systems or anyone looking to serve a model using vLLM. InstructLab provides Safetensor versions of our Granite models on HuggingFace.
```

   > **NOTE:** If multiple `ilab` clients try to connect to the same InstructLab server at the same time, the 1st will connect to the server while the others will start their own temporary server. This will require additional resources on the host machine.

> - Serve a non-default Safetensor model (e.g. Granite-7b-lab) NOTE: this requires a GPU.

Reviewer: Please add a dot between sentences

   > **NOTE:** If multiple `ilab` clients try to connect to the same InstructLab server at the same time, the 1st will connect to the server while the others will start their own temporary server. This will require additional resources on the host machine.

> - Serve a non-default Safetensors model (e.g. Granite-7b-lab). NOTE: this requires a GPU.

Reviewer 4: ```suggestion
- Serve a non-default Safetensors model (e.g. granite-7b-lab). NOTE: this requires a GPU.
```

ilab model chat --model models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf
```

> Please note that usage of `--model` necessitates that the existing server has that model. If not, you must exit the server. `--model` in `ilab chat` has the ability to start a server on your behalf with the specified model if one is not already running on the port.

Reviewer 4: ```suggestion
Please note that usage of `--model` necessitates that the existing server has that model. If not, you must exit the server. `--model` in `ilab model chat` has the ability to start a server on your behalf with the specified model if one is not already running on the port.
```

### 🚀 Generate a synthetic dataset

> Before following these instructions, ensure the existing model you are adding skills or knowledge to is still running. Alternatively, `ilab data generate` can start a server for you if you provide a fully qualified model path via `--model`

Reviewer: Please add a trailing dot

   ...
   ```

>    The synthetic data set will be two files in the newly created `~/.local/share/instructlab/datasets` directory named `skills_train_msgs_*.jsonl` and `knowledge_train_msgs_*.jsonl`

Reviewer: Add a trailing dot

>    The synthetic data set will be three files in the newly created `~/.local/share/instructlab/datasets` directory named `skills_train_msgs_*.jsonl` and `knowledge_train_msgs_*.jsonl`

Reviewer 3: ```suggestion
   The synthetic data set will be two files in the newly created `~/.local/share/instructlab/datasets` directory named `skills_train_msgs_*.jsonl` and `knowledge_train_msgs_*.jsonl`
```

> 2. Verify the files have been created by running the `ls datasets` command. Note: you must be in your XDG_DIR/instructlab directory.

Reviewer 3: May not be XDG_DIR on MacOS - in this case it would be `Library/Application\ Support/instructlab`

Reviewer: XDG_DATA_HOME? (need to check if the right variable - but definitely there's no envvar called XDG_DIR).

> 2. Verify the files have been created by running the `ls datasets` command. Note: you must be in your XDG_DATA_HOME/instructlab directory.

Reviewer 4: ```suggestion
2. Verify the files have been created by running the `ls datasets` command. Note: you must be in your `XDG_DATA_HOME/instructlab` directory.
```

> **NOTE:** ⏳ This step can potentially take **several hours** to complete depending on your computing resources. Please stop `ilab model chat` and `ilab model serve` first to free resources.

If you are using `ilab model train --legacy` or are on MacOS:

`ilab model train` outputs a brand-new model that can be served in the `models` directory called `ggml-model-f16.gguf`.

If you are using `ilab model train` with a GPU enabled system:

`ilab model train` outputs brand-new models that can be served in the `~/.local/share/instructlab/checkpoints` directory.  These models can be run through `ilab model evaluate` to choose the best one.

> If you are using `ilab model train --strategy lab-multiphase`

Reviewer: lost some text?

Author: the ls models is outdated and not very helpful so I removed it, can add it back if others disagree!

Reviewer 4: I think @booxter is referring to the fact this says 
```text
If you are using `ilab model train --strategy lab-multiphase`
```
followed by nothing, which makes no sense

```shell
ilab model train --device=cuda

```
This version of `ilab model train` outputs brand-new models that can be served in the `~/.local/share/instructlab/checkpoints` directory.  These models can be run through `ilab model evaluate` to choose the best one.

#### Train the model locally with multi-phase training and GPU acceleration

> `ilab model train` supports multi-phase training. This means in phase1 we train the model on knowledge, followed by evaluation which chooses the best checkpoint from that phase. The best checkpoint is then promoted to the starter model of phase2 which trains the model on skills. The skill phase is followed by evaluation in which the best checkpoint is chosen. Here is an example of how to run multi-phase training

Reviewer 3: Can reformat this to:
```suggestion
`ilab model train` supports multi-phase training. This results in the following workflow:
1. We train the model on knowledge
2. Evaluate the trained model to find the best checkpoint
3. We train the model on skills
4. We evaluate the model to find the best overall checkpoint
```

```shell

> ilab model train --strategy lab-multiphase --phased-phase1-data <KNOWLEDGE_TRAIN_MSGS_JSONL> --phased-phase2-data <SKILLS_TRAIN_MSGS_JSONL> -y`

Reviewer 3: I recommend formatting the dataset placeholders in a way where it's clear that it's a file placeholder:

```suggestion
ilab model train --strategy lab-multiphase --phased-phase1-data <knowledge train messages jsonl> --phased-phase2-data <skills train messages jsonl> -y`
```

```

> This command takes in two jsonl files from your datasets dir, one is the knowledge jsonl and the other is a skills jsonl. the -y skips and interactive prompt asking the user if they are sure they want to run multi-phase training.

Reviewer 3: ```suggestion
This command takes in two jsonl files from your datasets dir, one is the knowledge jsonl and the other is a skills jsonl. the `-y` flag skips and interactive prompt asking the user if they are sure they want to run multi-phase training.
```

> Note: this command will take 3hrs+ depending on the size of the data and number of training epochs you run.

Reviewer 3: ```suggestion
Note: this command may take 3 or more hours depending on the size of the data and number of training epochs you run.
```

```shell
ilab model train --device=cuda

```
This version of `ilab model train` outputs brand-new models that can be served in the `~/.local/share/instructlab/checkpoints` directory.  These models can be run through `ilab model evaluate` to choose the best one.

#### Train the model locally with multi-phase training and GPU acceleration

`ilab model train` supports multi-phase training. This results in the following workflow:
1. We train the model on knowledge
2. Evaluate the trained model to find the best checkpoint
3. We train the model on skills
4. We evaluate the model to find the best overall checkpoint

```shell

> ilab model train --strategy lab-multiphase --phased-phase1-data <knowledge train messages jsonl> --phased-phase2-data <skills train messages jsonl> -y`

Reviewer: remove "`" at the end?

```

> This command takes in two jsonl files from your datasets dir, one is the knowledge jsonl and the other is a skills jsonl. the `-y` flag skips and interactive prompt asking the user if they are sure they want to run multi-phase training.

Reviewer: dir -> directory? "and interactive prompt" -> "an interactive prompt"

ilab model train --device=cuda
```

This version of `ilab model train` outputs brand-new models that can be served in the `~/.local/share/instructlab/checkpoints` directory on Linux and `~/Library/Application Support/instructlab/checkpoints` on MacOS.  These models can be run through `ilab model evaluate` to choose the best one.

#### Train the model locally with multi-phase training and GPU acceleration

`ilab model train` supports multi-phase training. This results in the following workflow:

1. We train the model on knowledge
2. Evaluate the trained model to find the best checkpoint
3. We train the model on skills
4. We evaluate the model to find the best overall checkpoint

```shell
ilab model train --strategy lab-multiphase --phased-phase1-data <knowledge train messages jsonl> --phased-phase2-data <skills train messages jsonl> -y
```

> This command takes in two jsonl files from your datasets directory, one is the knowledge jsonl and the other is a skills jsonl. the `-y` flag skips an interactive prompt asking the user if they are sure they want to run multi-phase training.

Reviewer 4: ```suggestion
This command takes in two `.jsonl` files from your `datasets` directory, one is the knowledge jsonl and the other is a skills jsonl. The `-y` flag skips an interactive prompt asking the user if they are sure they want to run multi-phase training.
```
